File: results/dnd-rest-benchmark/lifecycle-runs/20260730T150315Z_pi_kimi-k2p7-code_typescript-node/src/routes/compendium.ts
import { ServerResponse } from 'node:http';
import { sendJSON } from '../http-utils.js';
import { createItem, createMonster, getItem, getMonster, itemExists, monsterExists } from '../repository.js';
import { isNonEmptyString, isStringArray } from '../validators.js';
import type { Item, Monster } from '../types.js';

export function handleGetMonster(res: ServerResponse, params: Record<string, string>): void {
  const monster = getMonster(params.slug);
  if (!monster) {
    sendJSON(res, 404, { error: 'not found' });
    return;
  }
  sendJSON(res, 200, monster);
}

export function handleCreateMonster(res: ServerResponse, _params: unknown, body: unknown): void {
  const { slug, name, cr, armor_class, hit_points, tags } = body as Record<string, unknown>;
  if (
    !isNonEmptyString(slug) ||
    typeof name !== 'string' ||
    typeof cr !== 'string' ||
    !Number.isInteger(armor_class) ||
    !Number.isInteger(hit_points)
  ) {
    sendJSON(res, 400, { error: 'invalid input' });
    return;
  }

  const parsedTags = tags === undefined ? [] : isStringArray(tags) ? tags : null;
  if (parsedTags === null) {
    sendJSON(res, 400, { error: 'invalid input' });
    return;
  }

  if (monsterExists(slug)) {
    sendJSON(res, 409, { error: 'slug already exists' });
    return;
  }

  const monster: Monster = {
    slug,
    name,
    cr,
    armor_class: Number(armor_class),
    hit_points: Number(hit_points),
    tags: parsedTags,
  };
  createMonster(monster);
  sendJSON(res, 201, {
    slug: monster.slug,
    name: monster.name,
    cr: monster.cr,
    armor_class: monster.armor_class,
    hit_points: monster.hit_points,
  });
}

export function handleGetItem(res: ServerResponse, params: Record<string, string>): void {
  const item = getItem(params.slug);
  if (!item) {
    sendJSON(res, 404, { error: 'not found' });
    return;
  }
  sendJSON(res, 200, item);
}

export function handleCreateItem(res: ServerResponse, _params: unknown, body: unknown): void {
  const { slug, name, type, rarity, cost_gp } = body as Record<string, unknown>;
  if (
    !isNonEmptyString(slug) ||
    typeof name !== 'string' ||
    typeof type !== 'string' ||
    typeof rarity !== 'string' ||
    !Number.isInteger(cost_gp)
  ) {
    sendJSON(res, 400, { error: 'invalid input' });
    return;
  }

  if (itemExists(slug)) {
    sendJSON(res, 409, { error: 'slug already exists' });
    return;
  }

  const item: Item = {
    slug,
    name,
    type,
    rarity,
    cost_gp: Number(cost_gp),
  };
  createItem(item);
  sendJSON(res, 201, item);
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075550Z_pi_kimi-k2p7-code_ruby-sinatra/lib/routes/compendium.rb
# frozen_string_literal: true

# Route definitions for compendium endpoints.

# --- Compendium ---

post '/v1/compendium/monsters' do
  content_type :json
  body = parse_json_body

  validate_monster_body!(body)

  json_error(409, 'monster already exists') if Storage.monster_exists?(body['slug'])

  Storage.create_monster(
    body['slug'],
    body['name'],
    body['cr'],
    body['armor_class'],
    body['hit_points'],
    body['tags']
  )

  status 201
  JSON.dump(
    slug: body['slug'],
    name: body['name'],
    cr: body['cr'],
    armor_class: body['armor_class'],
    hit_points: body['hit_points']
  )
end

get '/v1/compendium/monsters/:slug' do
  content_type :json
  monster = Storage.load_monster(params[:slug])
  json_error(404, 'monster not found') unless monster

  JSON.dump(monster)
end

post '/v1/compendium/items' do
  content_type :json
  body = parse_json_body

  validate_item_body!(body)

  json_error(409, 'item already exists') if Storage.item_exists?(body['slug'])

  Storage.create_item(
    body['slug'],
    body['name'],
    body['type'],
    body['rarity'],
    body['cost_gp']
  )

  status 201
  JSON.dump(
    slug: body['slug'],
    name: body['name'],
    type: body['type'],
    rarity: body['rarity'],
    cost_gp: body['cost_gp']
  )
end

get '/v1/compendium/items/:slug' do
  content_type :json
  item = Storage.load_item(params[:slug])
  json_error(404, 'item not found') unless item

  JSON.dump(item)
end

